### ISSUE TITLE
Modbus error

### ISSUE THREAD

**Issue by flopp999** [derenback/Domoticz-SMA-Inverter#14]
Suddenly this happened. It has been working perfect for many months.
Any ideas?

"
2022-08-13 18:30:09.819 [7f72a37fe700] Error: SMA modbus: 'onHeartbeat' failed 'ParameterException':'Modbus Error: [Invalid Parameter] Invalid collection of registers supplied'.
2022-08-13 18:30:09.819 [7f72a37fe700] Error: SMA modbus: Exception traceback:
2022-08-13 18:30:09.819 [7f72a37fe700] Error: SMA modbus: ----> Line 208 in '/home/debian/domoticz/plugins/Domoticz-SMA-Inverter/plugin.py'
2022-08-13 18:30:09.819 [7f72a37fe700] Error: SMA modbus: ----> Line 156 in '/home/debian/domoticz/plugins/Domoticz-SMA-Inverter/plugin.py'
2022-08-13 18:30:09.819 [7f72a37fe700] Error: SMA modbus: ----> Line 62 in '/home/debian/domoticz/plugins/Domoticz-SMA-Inverter/plugin.py'
2022-08-13 18:30:09.819 [7f72a37fe700] Error: SMA modbus: ----> Line 316 in '/usr/local/lib/python3.9/dist-packages/pymodbus/payload.py'
2022-08-13 18:30:39.947 [7f72a37fe700] Error: SMA modbus: 'onHeartbeat' failed 'ParameterException':'Modbus Error: [Invalid Parameter] Invalid collection of registers supplied'.
2022-08-13 18:30:39.947 [7f72a37fe700] Error: SMA modbus: Exception traceback:
2022-08-13 18:30:39.947 [7f72a37fe700] Error: SMA modbus: ----> Line 208 in '/home/debian/domoticz/plugins/Domoticz-SMA-Inverter/plugin.py'
2022-08-13 18:30:39.947 [7f72a37fe700] Error: SMA modbus: ----> Line 156 in '/home/debian/domoticz/plugins/Domoticz-SMA-Inverter/plugin.py'
2022-08-13 18:30:39.947 [7f72a37fe700] Error: SMA modbus: ----> Line 62 in '/home/debian/domoticz/plugins/Domoticz-SMA-Inverter/plugin.py'
2022-08-13 18:30:39.947 [7f72a37fe700] Error: SMA modbus: ----> Line 316 in '/usr/local/lib/python3.9/dist-packages/pymodbus/payload.py'
"

**Comment by derenback** [comment]
Have you updated your system? Python version/Domoticz version or something else? 
What version of the plugin are you using? 
Have you tried updating to the latest version if not already on it?

**Comment by derenback** [comment]
You can also try the new test script and see if you get any other information.
https://github.com/derenback/Domoticz-SMA-Inverter/blob/main/test/sma_test.py

**Comment by flopp999** [comment]
Latest Debian
Latest version of plugin 0.9.2
Python 3.9.2
DZ 2022.1

I restarted SMA and then it started to work.

I didn't have time to test manually but will do it next time so maybe there is a solution that can be implemented in the plugin

**Comment by derenback** [comment]
Hi! I have noticed issues similar to this and tried to fix it in version 0.9.4. The problem seems to be on the SMA side when a connection timeout on the network the port or socket is not released as it should on the SMA unit and it is no longer possible to connect. In this case I found the only easy way to fix it was to simply reboot the inverter. I also tried to turn off Modbus TCP on the admin page for installers and back on again. This also works... but I have not found a way to reset the connection from the plugin.

**Comment by derenback** [comment]
Closing this for now. If you still have issues please reopen the case. Looks like the restart of the inverter helped...

**Comment by flopp999** [comment]
I found another way to restart modbus on inverter.
Deactivate MODBUS in inverter, then active. It started to work for me.
Maybe auto_close as an argument can help, then the connection will be closed after each fetch?
https://pymodbustcp.readthedocs.io/en/latest/package/class_ModbusClient.html

**Comment by derenback** [comment]
That was what I also found as a workaround. (Disable and enable Modbus on the inverter). Unfortunately the auto_close functionality does not help with this and as of my testing it does not really work that well at all to do many requests a fast speed.
Since the issue seems to be connected to some sort of external issue that leads to a timeout on the Inverter side it does not help... not for me during my testing.

I tested it to start with by opening a connection. Request some data. Disconnected the ethernet cable or turned of a switch on the way between the computer and inverter. Wait some time. Try to reconnect. Connection times out until modbus is reset or the inverter is restarted.
